src/dataset-builder/ConceptSearch.test.ts

```diff
@@ -0,0 +1,136 @@
+import userEvent from '@testing-library/user-event';
+import { h } from 'react-hyperscript-helpers';
+import { ConceptSearch } from 'src/dataset-builder/ConceptSearch';
+import { dummyDatasetModel, dummyGetConceptForId } from 'src/dataset-builder/TestConstants';
+import { DataRepo, DataRepoContract, SnapshotBuilderConcept } from 'src/libs/ajax/DataRepo';
+import { asMockedFn, renderWithAppContexts as render } from 'src/testing/test-utils';
+
+jest.mock('src/libs/ajax/GoogleStorage');
+type DataRepoExports = typeof import('src/libs/ajax/DataRepo');
+jest.mock('src/libs/ajax/DataRepo', (): DataRepoExports => {
+  return {
+    ...jest.requireActual('src/libs/ajax/DataRepo'),
+    DataRepo: jest.fn(),
+  };
+});
+describe('ConceptSearch', () => {
+  const onCancel = jest.fn();
+  const onCommit = jest.fn();
+  const onOpenHierarchy = jest.fn();
+  const actionText = 'action text';
+  const datasetId = '0';
+  const initialCart: SnapshotBuilderConcept[] = [];
+  const domainOption = dummyDatasetModel()!.snapshotBuilderSettings!.domainOptions[0];
+  const renderSearch = () => {
+    render(
+      h(ConceptSearch, {
+        actionText,
+        initialSearch: '',
+        initialCart,
+        onCancel,
+        onCommit,
+        onOpenHierarchy,
+        datasetId,
+        domainOption,
+      })
+    );
+  };
+
+  const displayedConcepts = [dummyGetConceptForId(102), dummyGetConceptForId(103)];
+  const mockSearch = jest.fn();
+
+  const mockDataRepoContract: DataRepoContract = {
+    dataset: (_datasetId) =>
+      ({
+        searchConcepts: mockSearch,
+      } as Partial<DataRepoContract['dataset']>),
+  } as Partial<DataRepoContract> as DataRepoContract;
+  asMockedFn(DataRepo).mockImplementation(() => mockDataRepoContract as DataRepoContract);
+  mockSearch.mockResolvedValue({
+    result: displayedConcepts,
+  });
+
+  it('renders the concept search with an initial search result', async () => {
+    // Arrange
+    renderSearch();
+    // Assert
+    expect(await screen.findByText(displayedConcepts[0].name)).toBeTruthy();
+    expect(await screen.findByText(displayedConcepts[0].id)).toBeTruthy();
+    expect(await screen.findByText(displayedConcepts[1].name)).toBeTruthy();
+    expect(await screen.findByText(displayedConcepts[1].id)).toBeTruthy();
+    // Action text not visible until a row is selected.
+    expect(screen.queryByText(actionText)).toBeFalsy();
+  });
+
+  it('filters the view based on user input', async () => {
+    // Arrange
+    renderSearch();
+    const searchText = displayedConcepts[0].name;
+    // Act
+    const user = userEvent.setup();
+    await user.type(screen.getByPlaceholderText('Search'), searchText);
+    // Assert - the search is called with the search text, after debounce.
+    await new Promise((r) => setTimeout(r, 1000));
+    expect(mockSearch).toHaveBeenCalledWith(domainOption.root, searchText);
+    expect(screen.findByText(displayedConcepts[0].name)).toBeTruthy();
```

Comment:
What you can do to avoid this await is move the `screen.findByText` above, and then await on the `findByText`. Because `findByText` is async friendly it will wait for the text to be visible. If isn't ever visible, then it will error.

Reply:
I changed the test to verify that the TDR API is being called rather than checking the UI contents. Also the test now mocks out the debounce code so it no longer needs to await a dummy promise.